File: src/jiggler.py
"""jiggler.py — Async mouse jiggler.

Sends small, randomized mouse movements at configurable intervals
to prevent screen lock / idle detection.
"""

from __future__ import annotations

import asyncio
import logging
import random
import struct

from src.config import JigglerConfig
from src.hid_transport import HIDTransport

log = logging.getLogger(__name__)


def _mouse_report(buttons: int, dx: int, dy: int, wheel: int = 0) -> bytes:
    """Build a 4-byte mouse HID report: [buttons, dx(int8), dy(int8), wheel(int8)]."""
    dx = max(-127, min(127, dx))
    dy = max(-127, min(127, dy))
    wheel = max(-127, min(127, wheel))
    return struct.pack("Bbbb", buttons, dx, dy, wheel)


class MouseJiggler:
    """Async mouse jiggler that sends tiny movements via HID transport."""

    def __init__(self, transport: HIDTransport, cfg: JigglerConfig) -> None:
        self.transport = transport
        self.cfg = cfg
        self._task: asyncio.Task | None = None

    async def run(self) -> None:
        """Main jiggle loop — runs until cancelled."""
        if not self.cfg.enabled:
            log.info("Jiggler disabled in config")
            return

        log.info(
            "Jiggler started (interval %d–%ds, max_delta %d)",
            self.cfg.interval_min,
            self.cfg.interval_max,
            self.cfg.max_delta,
        )

        try:
            while True:
                interval = random.uniform(self.cfg.interval_min, self.cfg.interval_max)
                log.debug("Next jiggle in %.1fs", interval)
                await asyncio.sleep(interval)

                try:
                    await self._jiggle_once()
                except Exception as e:
                    log.warning("Jiggle failed: %s", e)
        except asyncio.CancelledError:
            log.info("Jiggler stopped")

    async def _jiggle_once(self) -> None:
        """Send a small movement then approximately return to origin."""
        dx, dy = self._random_delta()
        log.debug("Jiggle: dx=%d dy=%d", dx, dy)

        # Move
        self.transport.send_mouse_report(_mouse_report(0, dx, dy))
        await asyncio.sleep(random.uniform(0.08, 0.2))

        # Return (slight offset so it's not perfectly zero)
        ret_x = -dx + random.choice([-1, 0, 1])
        ret_y = -dy + random.choice([-1, 0, 1])
        self.transport.send_mouse_report(_mouse_report(0, ret_x, ret_y))

    def _random_delta(self) -> tuple[int, int]:
        """Generate a random movement delta."""
        if random.random() < self.cfg.big_move_chance:
            d = random.randint(10, 20)
        else:
            d = self.cfg.max_delta

        dx = random.randint(-d, d)
        dy = random.randint(-d, d)

        # Avoid zero-movement
        if dx == 0 and dy == 0:
            dx = 1

        return dx, dy

    def start(self) -> asyncio.Task:
        """Create and return the jiggler asyncio task."""
        self._task = asyncio.create_task(self.run(), name="jiggler")
        return self._task

    def stop(self) -> None:
        """Cancel the jiggler task."""
        if self._task and not self._task.done():
            self._task.cancel()

File: rust/src/jiggler.rs
use anyhow::Result;
use log::{debug, info, warn};
use rand::Rng;
use std::sync::Arc;
use tokio::sync::watch;
use tokio::time::{sleep, Duration};

use crate::config::JigglerConfig;
use crate::transport::{self, HidTransport};

/// Async mouse jiggler loop.
pub async fn run<T: HidTransport>(
    transport: Arc<T>,
    cfg: JigglerConfig,
    mut stop: watch::Receiver<bool>,
) -> Result<()> {
    if !cfg.enabled {
        info!("Jiggler disabled in config");
        return Ok(());
    }

    info!("Jiggler started");
    let mut rng = rand::thread_rng();

    loop {
        // Random interval between jiggles
        let interval = rng.gen_range(cfg.interval_min..=cfg.interval_max);
        debug!("Next jiggle in {:.1}s", interval);

        // Interruptible sleep in 1-second chunks
        let deadline = tokio::time::Instant::now() + Duration::from_secs_f64(interval);
        loop {
            tokio::select! {
                _ = sleep(Duration::from_secs(1).min(deadline - tokio::time::Instant::now())) => {
                    if tokio::time::Instant::now() >= deadline {
                        break;
                    }
                }
                _ = stop.changed() => {
                    if *stop.borrow() {
                        info!("Jiggler stopped");
                        return Ok(());
                    }
                }
            }
        }

        // Check stop again before jiggling
        if *stop.borrow() {
            info!("Jiggler stopped");
            return Ok(());
        }

        if let Err(e) = jiggle_once(transport.as_ref(), &cfg).await {
            warn!("Jiggle failed: {}", e);
        }
    }
}

async fn jiggle_once<T: HidTransport>(transport: &T, cfg: &JigglerConfig) -> Result<()> {
    let mut rng = rand::thread_rng();

    // Decide movement magnitude
    let d = if rng.gen::<f64>() < cfg.big_move_chance {
        rng.gen_range(10..=20)
    } else {
        cfg.max_delta
    };

    let mut dx: i32 = rng.gen_range(-d..=d);
    let mut dy: i32 = rng.gen_range(-d..=d);

    // Avoid zero movement
    if dx == 0 && dy == 0 {
        dx = 1;
    }

    debug!("Jiggle: dx={} dy={}", dx, dy);

    // Move
    transport::send_mouse(
        transport,
        dx.clamp(-127, 127) as i8,
        dy.clamp(-127, 127) as i8,
        0,
        0,
    )
    .await?;

    // Brief pause
    let pause_ms = rng.gen_range(80..=200);
    sleep(Duration::from_millis(pause_ms)).await;

    // Return approximately to origin (small offset for naturalness)
    let ret_x = -dx + rng.gen_range(-1..=1);
    let ret_y = -dy + rng.gen_range(-1..=1);
    transport::send_mouse(
        transport,
        ret_x.clamp(-127, 127) as i8,
        ret_y.clamp(-127, 127) as i8,
        0,
        0,
    )
    .await?;

    Ok(())
}
